internal/httputil/certpoolwatcher.go

```diff
@@ -0,0 +1,108 @@
+package httputil
+
+import (
+	"crypto/x509"
+	"fmt"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/go-logr/logr"
+)
+
+type CertPoolWatcher struct {
+	generation int
+	dir        string
+	mx         sync.RWMutex
+	pool       *x509.CertPool
+	log        logr.Logger
+	watcher    *fsnotify.Watcher
+	done       chan bool
+}
+
+// Returns the current CertPool and the generation number
+func (cpw *CertPoolWatcher) Get() (*x509.CertPool, int, error) {
+	cpw.mx.RLock()
+	defer cpw.mx.RUnlock()
+	if cpw.pool == nil {
+		return nil, 0, fmt.Errorf("no certificate pool available")
+	}
+	return cpw.pool.Clone(), cpw.generation, nil
+}
+
+func (cpw *CertPoolWatcher) Done() {
+	cpw.done <- true
+}
+
+func NewCertPoolWatcher(caDir string, log logr.Logger) (*CertPoolWatcher, error) {
+	pool, err := NewCertPool(caDir, log)
+	if err != nil {
+		return nil, err
+	}
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, err
+	}
+	if err = watcher.Add(caDir); err != nil {
+		return nil, err
+	}
+
+	cpw := &CertPoolWatcher{
+		generation: 1,
+		dir:        caDir,
+		pool:       pool,
+		log:        log,
+		watcher:    watcher,
+		done:       make(chan bool),
+	}
+	go func() {
+		for {
+			select {
+			case <-watcher.Events:
+				cpw.drainEvents()
+				cpw.update()
+			case err := <-watcher.Errors:
+				log.Error(err, "error watching certificate dir")
+				os.Exit(1)
+			case <-cpw.done:
+				err := watcher.Close()
+				if err != nil {
+					log.Error(err, "error closing watcher")
```

Comment:
What happens with the watcher in the case that the close fails? I see the goroutine for handling events exits, but does the watcher continue to stick around filling the event queue and consuming memory?

Reply:
Since the `cpw` is still likely referenced _somewhere_ the memory for the notifier is still around; but the notify OS resources have been cleaned up the _best that they can be_.